controllers/hardwareclassification_controller.go

```diff
@@ -0,0 +1,172 @@
+// +kubebuilder:rbac:groups=metal3.io.sigs.k8s.io,resources=hardwareclassifications,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=metal3.io.sigs.k8s.io,resources=hardwareclassifications/status,verbs=get;update;patch
+func (r *HardwareClassificationReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
+	ctx := context.Background()
+
+	// Get HardwareClassificationController to get values for Namespace and ExpectedHardwareConfiguration
+	hardwareClassification := &hwcc.HardwareClassification{}
+
+	if err := r.Client.Get(ctx, req.NamespacedName, hardwareClassification); err != nil {
+		if apierrors.IsNotFound(err) {
+			return ctrl.Result{}, nil
+		}
+		return ctrl.Result{}, err
+	}
+
+	// Get ExpectedHardwareConfiguraton from hardwareClassification
+	extractedProfile := hardwareClassification.Spec.ExpectedHardwareConfiguration
+
+	// Checking Disk Count present or not if Disk individual size is present
+	if extractedProfile.Disk != nil {
+		if err := checkDiskCount(extractedProfile); err != nil {
+			r.Log.Error(nil, "Disk count error", "Error", err.Error())
+			return ctrl.Result{}, nil
+		}
+	}
+
+	r.Log.Info("Extracted hardware configurations successfully", "Profile", extractedProfile)
+
+	// fetch BMH list from BMO
+	validHostList := fetchBmhHostList(ctx, r, hardwareClassification.ObjectMeta.Namespace)
+
+	if len(validHostList) == 0 {
+		err := errors.New("No BareMetal Host found in ready state")
+		r.Log.Error(err, "Error Occurred")
+		return ctrl.Result{}, nil
+	}
+
+	// Extract introspection data for each configuration provided in profile
+	extractedHardwareDetails := extractHardwareDetails(extractedProfile, validHostList)
+
+	r.Log.Info("Extracted hardware introspection details successfully", "IntrospectionDetails", extractedHardwareDetails)
+
+	return ctrl.Result{}, nil
+}
+
+// fetchBmhHostList this function will fetch and return baremetal hosts in ready state
+func fetchBmhHostList(ctx context.Context, r *HardwareClassificationReconciler, namespace string) []bmh.BareMetalHost {
+
+	bmhHostList := bmh.BareMetalHostList{}
+	var validHostList []bmh.BareMetalHost
+
+	opts := &client.ListOptions{
+		Namespace: namespace,
+	}
+
+	// Get list of BareMetalHost from BMO
+	err := r.Client.List(ctx, &bmhHostList, opts)
+	if err != nil {
+		r.Log.Error(nil, "Unable to extract details", "error", err.Error())
+		return validHostList
+	}
+
+	// Get hosts in ready status from bmhHostList
+	for _, host := range bmhHostList.Items {
+		if host.Status.Provisioning.State == "ready" {
+			validHostList = append(validHostList, host)
+		}
+	}
+
+	return validHostList
+}
+
+// extractHardwareDetails this function will return map containing
+// introspection details for a host.
+func extractHardwareDetails(extractedProfile hwcc.ExpectedHardwareConfiguration,
+	bmhList []bmh.BareMetalHost) map[string]map[string]interface{} {
+
+	extractedHardwareDetails := make(map[string]map[string]interface{})
```

Comment:
It looks like we probably want a type for this to give us the benefit of type checking for individual fields instead of mapping to an interface type.